internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	requestStart = `
+{
+  "streams": [
+    {
+      "stream": {
+        "service": %q
+      },
+      "values": [`
+
+	requestEnd = `
+      ]
+    }
+  ]
+}`
+
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient implements logClient
+// We send a new request to Loki when either:
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	service     string
+	lokiURL     *url.URL
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
+	buffer *bytes.Buffer
+	// how many messages have been written so far
+	messages int
+	// write deadline for flushing buffer
+	deadline *time.Time
+}
+
+func newLokiClient(service, url string, maxMessages int, delay time.Duration) (*lokiClient, error) {
+	lokiURL, err := parseURL(url)
+	if err != nil {
+		return nil, err
+	}
+	c := &lokiClient{
+		service:     service,
+		lokiURL:     lokiURL,
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	c.reset()
+	return c, nil
+}
+
+func parseURL(rawURL string) (*url.URL, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + rawURL)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid Loki server address: %v", err)
+	}
+
+	// check for and set loki defaults if omitted from address
+	if u.Scheme == "" {
+		u.Scheme = "http"
+		if u.Port() == "" {
```

Comment:
maybe. I think it's nice to put in sensible defaults cause I'm lazy. If it's gonna cause more problems than it's worth though, we can just require a fully qualified URL

Reply:
Do you think we should disallow this?
```
log-targets:
  loki:
    location: 10.1.77.237
```
To me it seems reasonable to fill in defaults, so that the actual send URL becomes `http://10.1.77.237:3100`

Reply:
In this sense yes. As what we are asking for is a location and 101.77.237 doesn't describe a full location. Sure we can try and get smart and help the user out but are we helping them by adding http and 3100 when that is not the location and we just end up throwing errors further down in the code and making the user scratch their head even more.

In saying that I can see it from both sides of the conversation and I am happy to go with the flow. If I think of myself as a user I would rather just be explicit and supply the information up front to the program. I am only going to type it once if that. Saving 12 keystrokes is not the end of the world for myself v trawling through log messages to figure out why something has broken because things are being inferred on my behalf.

Reply:
Talked to Ben and we'll just require the full location. It's easier to be strict and relax this later rather than go the other way